rpcclient/infrastructure.go

```diff
@@ -1733,28 +1740,38 @@ func (c *Client) Send() error {
 		return nil
 	}
 
-	// clear batchlist in case of an error
-	defer func() {
+	batchResp, err := c.sendAsync().Receive()
+	if err != nil {
+		// Clear batchlist in case of an error.
+		//
+		// TODO(yy): need to double check to make sure there's no
+		// concurrent access to this batch list, otherwise we may miss
+		// some batched requests.
 		c.batchList = list.New()
-	}()
 
-	result, err := c.sendAsync().Receive()
-
-	if err != nil {
 		return err
 	}
 
-	for iter := c.batchList.Front(); iter != nil; iter = iter.Next() {
-		var requestError error
-		request := iter.Value.(*jsonRequest)
-		individualResult := result[request.id]
-		fullResult, err := json.Marshal(individualResult.Result)
+	// Iterate each response and send it to the corresponding request.
+	for id, resp := range batchResp {
+		// Perform a GC on batchList and requestMap before moving
+		// forward.
+		request := c.removeRequest(id)
+
+		// If there's an error, we log it and continue to the next
+		// request.
+		fullResult, err := json.Marshal(resp.Result)
 		if err != nil {
-			return err
+			log.Errorf("Unable to marshal result: %v for req=%v",
+				err, request.id)
+
```

Comment:
Since it's a list of requests batched together, if one or more requests failed, it should not affect sending responses to other requests. And nope, I think we should almost never get an unmarshal error.

Reply:
Yeah, an unmarshall request here would mean that the full node sent over garbled json.